src/backend/probe/common.rs

```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::backend::Datum;
+
+use std::{ffi::CStr, mem::MaybeUninit, os::raw::c_char, str::Utf8Error};
+
+use errno::errno;
+
+use libc::{uname, utsname};
+
+pub fn system_info() -> Datum {
+    fn system_info_string(mut utsname: utsname) -> Result<String, Utf8Error> {
+        fn array_to_str(array: &'_ mut [c_char; 65]) -> Result<&'_ str, Utf8Error> {
+            unsafe { CStr::from_ptr(array.as_mut_ptr()) }.to_str()
+        }
+
+        Ok(format!(
+            "{} {} {} {} {} {}",
+            array_to_str(&mut utsname.sysname)?,
+            array_to_str(&mut utsname.nodename)?,
+            array_to_str(&mut utsname.release)?,
+            array_to_str(&mut utsname.version)?,
+            array_to_str(&mut utsname.machine)?,
+            array_to_str(&mut utsname.domainname)?,
+        ))
+    }
+
+    let mut utsname = MaybeUninit::uninit();
+    // If uname returns non-zero then an error occurred.
+    let pass = unsafe { uname(utsname.as_mut_ptr()) } == 0;
+
+    Datum {
+        name: "System Info".to_string(),
+        pass,
+        info: Some(if !pass {
+            // Retrieve the last errno value and display it
+            // as a human-readable string.
+            format!("[{}]", errno())
+        } else {
+            system_info_string(unsafe { utsname.assume_init() })
+                .unwrap_or_else(|e| format!("[utf8 error: {}]", e))
+        }),
```

Comment:
nit; How about something like this? That `Some` constructor at the beginning is a bit confusing IMO and `pass` is redundant
```suggestion
        info: if unsafe { uname(utsname.as_mut_ptr()) } != 0 {
            // Retrieve the last errno value and display it
            // as a human-readable string.
            Some(format!("[{}]", errno()))
        } else {
            Some(system_info_string(unsafe { utsname.assume_init() })
                .unwrap_or_else(|e| format!("[utf8 error: {}]", e)))
        }),
```